lib/plugins.ts

```diff
@@ -18,16 +18,217 @@ export interface PluginConfig {
+
+  /**
+   * Between every webdriver action, Protractor calls browser.waitForAngular() to
+   * make sure that Angular has no outstanding $http or $timeout calls.
+   * You can use waitForCondition() to have Protractor additionally wait for your
+   * custom condition to be truthy.
+   *
+   * @this {Object} bound to module.exports
+   *
+   * @throws {*} If this function throws an error, a failed assertion is added to
+   *     the test results.
+   *
+   * @return {q.Promise<boolean>|boolean} If truthy, Protractor will continue onto
+   *     the next command.  If falsy, webdriver will continuously re-run this
+   *     function until it is truthy.  If a rejected promise is returned, a failed
+   *     assertion is added to the test results, and protractor will continue onto
+   *     the next command.
+   */
+  waitForCondition?: () => q.Promise<any>;
+
+  /**
+   * Used to turn off default checks for angular stability
+   *
+   * Normally Protractor waits for all $timeout and $http calls to be processed
+   * before executing the next command.  This can be disabled using
+   * browser.ignoreSynchronization, but that will also disable any
+   * <Plugin>.waitForPromise or <Plugin>.waitForCondition checks.  If you want
+   * to
+   * disable synchronization with angular, but leave in tact any custom plugin
+   * synchronization, this is the option for you.
+   *
+   * This is used by users who want to replace Protractor's synchronization code
+   * This is used by users who want to replace Protractor's synchronization code
+   * with their own.
+   *
+   * @type {boolean}
+   */
+  skipAngularStability?: boolean;
+
+  /**
+   * Used when reporting results.
+   *
+   * If you do not specify this property, it will be filled in with something
+   * reasonable (e.g. the plugin's path)
+   *
+   * @type {string}
+   */
+  name?: string;
+
+  /**
+   * The plugin configuration object. Note that this is not the entire
+   * Protractor config object, just the entry in the plugins array for this
+   * plugin.
+   *
+   * @type {Object}
+   */
+  config?: PluginConfig;
+
+  /**
+   * Adds a failed assertion to the test's results.
+   *
+   * @param {string} message The error message for the failed assertion
+   * @param {specName: string, stackTrace: string} options Some optional extra
+   *     information about the assertion:
+   *       - specName The name of the spec which this assertion belongs to.
+   *            Defaults to `PLUGIN_NAME + ' Plugin Tests'`.
+   *       - stackTrace The stack trace for the failure.  Defaults to undefined.
+   *     Defaults to `{}`.
+   *
+   * @throws {Error} Throws an error if called after results have been reported
+   */
+  addFailure?:
       (message?: string,
        info?: {specName?: string, stackTrace?: string}) => void;
-  addSuccess: (info?: {specName?: string}) => void;
-  addWarning: (message?: string, info?: {specName?: string}) => void;
+
+  /**
+   * Adds a passed assertion to the test's results.
```

Comment:
Mention that this is added by the protractor API - not to be implemented by the plugin author